internal/k8s/service/resolver.go

```diff
@@ -137,23 +105,23 @@ func (r *ResolutionErrors) Flatten() (ServiceResolutionErrorType, error) {
 		return NoResolutionErrorType, nil
 	}
 
-	if len(r.genericErrors) != 0 || (len(r.consulErrors) != 0 && len(r.k8sErrors) != 0 && len(r.refNotPermittedErrors) != 0) {
+	//return generic error if there are multiple errors types, or if generic errors exist
+	if len(r.errors[GenericResolutionErrorType]) != 0 || (len(r.errors) > 1) {
 		return GenericResolutionErrorType, errors.New(r.String())
 	}
 
-	if len(r.consulErrors) != 0 {
-		return ConsulServiceResolutionErrorType, errors.New(r.String())
-	}
-
-	if len(r.refNotPermittedErrors) != 0 {
-		return RefNotPermittedErrorType, errors.New(r.String())
+	//otherwise assume that the first error type in the map is the error we found
+	for errType := range r.errors {
```

Comment:
so, Go map iteration order [isn't meant to be stable](https://go.dev/blog/maps#iteration-order):

> When iterating over a map with a range loop, the iteration order is not specified and is not guaranteed to be the same from one iteration to the next.

In practice it's actually [based off of hash keys and collision indexes](https://github.com/golang/go/blob/19309779ac5e2f5a2fd3cbb34421dafb2855ac21/src/runtime/map.go#L23-L31). So, if we're trying to do any sort of prioritization here, we'll need to specify an order via something like:

```go
// this array determines the priority of the error types returned
// when the resolution errors are flattened
var orderedResolutionErrorTypes = []ServiceResolutionErrorType{
  K8sServiceResolutionErrorType,
  ConsulServiceResolutionErrorType,
  RefNotPermittedErrorType,
  GenericResolutionErrorType,
}

...

func (r *ResolutionErrors) Flatten() (ServiceResolutionErrorType, error) {
  if r.Empty() {
    return NoResolutionErrorType, nil
  }

  for _, errorType := range orderedResolutionErrorTypes {
    if _, found := r.refNotPermittedErrors[errorType]; found {
      return errType, errors.New(r.String())
    }
  }

  // should not be reached
  return NoResolutionErrorType, nil
}
```

You can add the ordering array right at the top near the type specification just to make it so you don't have to track things down across the file.